src/native/crashtracker/crashtracker_runtime_stacks.rs

```diff
@@ -0,0 +1,341 @@
+            }
+        }
+        emit_stacktrace_string(buffer.as_ptr() as *const c_char);
+        return;
+    }
+
+    emit_stacktrace_string("<traceback_read_failed>\0".as_ptr() as *const c_char);
+}
+
+pub unsafe extern "C" fn native_runtime_stack_string_callback(
+    emit_stacktrace_string: unsafe extern "C" fn(*const c_char),
+) {
+    dump_python_traceback_as_string(emit_stacktrace_string);
+}
+
+/************************************************************
+ Emit runtime stacktrace as frames using by traversing the frame object chain
+************************************************************/
+
+#[cfg(unix)]
+const FRAME_FUNCTION_CAP: usize = 256;
+#[cfg(unix)]
+const FRAME_FILE_CAP: usize = 512;
+
+#[cfg(unix)]
+static FRAME_COLLECTION_GUARD: AtomicBool = AtomicBool::new(false);
+
+#[cfg(unix)]
+unsafe fn capture_frames_via_python(emit_frame: unsafe extern "C" fn(&RuntimeStackFrame)) {
+    let current = pyo3_ffi::PyThreadState_Get();
+    if !current.is_null() {
+        let _ = collect_and_emit_frames_for_thread(current, emit_frame);
+    }
+}
+
+#[cfg(unix)]
+unsafe fn collect_and_emit_frames_for_thread(
+    tstate: *mut pyo3_ffi::PyThreadState,
+    emit_frame: unsafe extern "C" fn(&RuntimeStackFrame),
+) -> bool {
+    if tstate.is_null() {
+        return false;
+    }
+
+    let mut emitted = false;
+    let mut frame = thread_top_frame(tstate);
+
+    while !frame.is_null() {
+        if emit_python_frame(frame, emit_frame) {
+            emitted = true;
+        }
+        frame = advance_frame(frame);
+    }
+
+    emitted
+}
+
+unsafe fn thread_top_frame(tstate: *mut pyo3_ffi::PyThreadState) -> *mut pyo3_ffi::PyFrameObject {
+    if tstate.is_null() {
+        ptr::null_mut()
+    } else {
+        // We won't support Python 3.9
+        #[cfg(not(Py_3_10))]
+        let frame: *mut pyo3_ffi::PyFrameObject = ptr::null_mut();
+
+        #[cfg(Py_3_10)]
+        let frame = pyo3_ffi::PyThreadState_GetFrame(tstate);
+
+        #[cfg(not(Py_3_11))]
+        {
+            if !frame.is_null() {
+                // On CPython <= 3.10 it returns/holds a borrowed frame, so we need to
+                // Py_XINCREF to keep it alive while we traverse.
+                pyo3_ffi::Py_XINCREF(frame as *mut pyo3_ffi::PyObject);
+            }
+        }
+        frame
+    }
+}
+
+unsafe fn advance_frame(frame: *mut pyo3_ffi::PyFrameObject) -> *mut pyo3_ffi::PyFrameObject {
+    if frame.is_null() {
+        return ptr::null_mut();
+    }
+    let back = pyo3_ffi::PyFrame_GetBack(frame);
+    pyo3_ffi::Py_DecRef(frame as *mut pyo3_ffi::PyObject);
+    back
+}
+
+#[cfg(unix)]
+unsafe fn emit_python_frame(
+    frame: *mut pyo3_ffi::PyFrameObject,
+    emit_frame: unsafe extern "C" fn(&RuntimeStackFrame),
+) -> bool {
+    if frame.is_null() {
+        return false;
+    }
+
+    let file_view = get_code_attr_utf8_view(frame, b"co_filename\0");
+    let function_view = get_code_attr_utf8_view(frame, b"co_name\0");
```

Comment:
I'd prefer using `co_qualname` similar to what we do in the profiler for Python 3.11+

https://github.com/DataDog/dd-trace-py/pull/15436